File: leon-core/server/src/core/paul/temporal.ts
import type { PaulDateRange } from './types'

const WEEKDAYS: Record<string, number> = {
  dimanche: 0,
  dimanchee: 0,
  lundi: 1,
  mardi: 2,
  mercredi: 3,
  jeudi: 4,
  vendredi: 5,
  samedi: 6
}

const MONTHS: Record<string, number> = {
  janvier: 0,
  fevrier: 1,
  février: 1,
  mars: 2,
  avril: 3,
  mai: 4,
  juin: 5,
  juillet: 6,
  aout: 7,
  août: 7,
  septembre: 8,
  octobre: 9,
  novembre: 10,
  decembre: 11,
  décembre: 11
}

function normalize(value: string): string {
  return value
    .toLocaleLowerCase('fr-FR')
    .normalize('NFD')
    .replace(/[\u0300-\u036f]/g, '')
    .replace(/\s+/g, ' ')
    .trim()
}

function pad(value: number): string {
  return String(value).padStart(2, '0')
}

function dayKey(date: Date): string {
  return `${date.getUTCFullYear()}-${pad(date.getUTCMonth() + 1)}-${pad(date.getUTCDate())}`
}

function parseDayKey(value: string): Date {
  return new Date(`${value}T00:00:00.000Z`)
}

function calendarParts(date: Date, timezone: string): { year: number, month: number, day: number, hour: number, minute: number, second: number } {
  const parts = new Intl.DateTimeFormat('en-CA', {
    timeZone: timezone,
    hourCycle: 'h23',
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
    hour: '2-digit',
    minute: '2-digit',
    second: '2-digit'
  }).formatToParts(date)
  const values = Object.fromEntries(parts.map((part) => [part.type, part.value]))
  return {
    year: Number(values['year']),
    month: Number(values['month']),
    day: Number(values['day']),
    hour: Number(values['hour']),
    minute: Number(values['minute']),
    second: Number(values['second'])
  }
}

function zonedMidnight(day: string, timezone: string): Date {
  const calendar = parseDayKey(day)
  const guess = calendar.getTime()
  const represented = calendarParts(new Date(guess), timezone)
  const representedAsUtc = Date.UTC(represented.year, represented.month - 1, represented.day, represented.hour, represented.minute, represented.second)
  const firstPass = guess - (representedAsUtc - guess)
  const corrected = calendarParts(new Date(firstPass), timezone)
  const correctedAsUtc = Date.UTC(corrected.year, corrected.month - 1, corrected.day, corrected.hour, corrected.minute, corrected.second)
  return new Date(guess - (correctedAsUtc - firstPass))
}

export function dayKeyInTimezone(timestamp: number, timezone = process.env['PAUL_TIMEZONE'] || 'Europe/Paris'): string {
  const parts = calendarParts(new Date(timestamp), timezone)
  return `${parts.year}-${pad(parts.month)}-${pad(parts.day)}`
}

function addDays(date: Date, amount: number): Date {
  const result = new Date(date.getTime())
  result.setUTCDate(result.getUTCDate() + amount)
  return result
}

function startOfWeek(date: Date): Date {
  const result = new Date(date.getTime())
  const mondayOffset = (result.getUTCDay() + 6) % 7
  result.setUTCDate(result.getUTCDate() - mondayOffset)
  return result
}

function toRange(start: Date, end: Date, expression: string | null, timezone: string): PaulDateRange {
  const startDay = dayKey(start)
  const endDay = dayKey(end)
  const startAt = zonedMidnight(startDay, timezone).getTime()
  const nextDay = dayKey(addDays(parseDayKey(endDay), 1))
  const endAt = zonedMidnight(nextDay, timezone).getTime() - 1

  return {
    startDay,
    endDay,
    startAt,
    endAt,
    expression
  }
}

function resolveWeekday(
  target: number,
  now: Date,
  direction: 'past' | 'future' | 'nearest'
): Date {
  const current = now.getUTCDay()
  let delta = target - current

  if (direction === 'past' && delta >= 0) {
    delta -= 7
  }
  if (direction === 'future' && delta <= 0) {
    delta += 7
  }
  if (direction === 'nearest' && Math.abs(delta) > 3) {
    delta += delta > 0 ? -7 : 7
  }

  return addDays(now, delta)
}

function parseExplicitDate(expression: string, now: Date): Date | null {
  const numeric = expression.match(/\b(\d{1,2})[/-](\d{1,2})(?:[/-](\d{2,4}))?\b/)
  if (numeric) {
    const day = Number(numeric[1])
    const month = Number(numeric[2]) - 1
    const rawYear = numeric[3]
    const year = rawYear
      ? rawYear.length === 2
        ? 2000 + Number(rawYear)
        : Number(rawYear)
      : now.getUTCFullYear()
    if (day >= 1 && day <= 31 && month >= 0 && month <= 11) {
      return new Date(Date.UTC(year, month, day))
    }
  }

  const written = normalize(expression).match(
    /\b(\d{1,2})\s+(?:de\s+)?([a-zéû]+)(?:\s+(\d{4}))?\b/
  )
  if (written) {
    const month = MONTHS[written[2] || '']
    if (month !== undefined) {
      return new Date(Date.UTC(
        written[3] ? Number(written[3]) : now.getUTCFullYear(),
        month,
        Number(written[1])
      ))
    }
  }

  return null
}

export function resolveDateRange(
  expression: string | null | undefined,
  now = new Date(),
  direction: 'past' | 'future' | 'nearest' = 'nearest',
  timezone = process.env['PAUL_TIMEZONE'] || 'Europe/Paris'
): PaulDateRange {
  const raw = String(expression || '').trim()
  const value = normalize(raw)
  const todayParts = calendarParts(now, timezone)
  const today = new Date(Date.UTC(todayParts.year, todayParts.month - 1, todayParts.day))

  if (!value || value === 'aujourd\'hui' || value === 'aujourd hui' || value === 'ce jour') {
    return toRange(today, today, raw || null, timezone)
  }
  if (value === 'demain' || value === 'le lendemain') {
    const date = addDays(today, 1)
    return toRange(date, date, raw, timezone)
  }
  if (value === 'hier' || value === 'la veille') {
    const date = addDays(today, -1)
    return toRange(date, date, raw, timezone)
  }

  const relative = value.match(/(?:dans|d'ici|d ici)\s+(\d+)\s+jours?/) 
  if (relative) {
    const date = addDays(today, Number(relative[1]))
    return toRange(date, date, raw, timezone)
  }

  if (value.includes('cette semaine') || value === 'la semaine') {
    const start = startOfWeek(today)
    return toRange(start, addDays(start, 6), raw, timezone)
  }
  if (value.includes('semaine derniere')) {
    const end = addDays(startOfWeek(today), -1)
    return toRange(addDays(end, -6), end, raw, timezone)
  }
  if (value.includes('semaine prochaine')) {
    const start = addDays(startOfWeek(today), 7)
    return toRange(start, addDays(start, 6), raw, timezone)
  }
  if (value.includes('ce mois') || value.includes('ce mois-ci')) {
    const start = new Date(Date.UTC(today.getUTCFullYear(), today.getUTCMonth(), 1))
    const end = new Date(Date.UTC(today.getUTCFullYear(), today.getUTCMonth() + 1, 0))
    return toRange(start, end, raw, timezone)
  }

  const explicit = parseExplicitDate(value, today)
  if (explicit) {
    return toRange(explicit, explicit, raw, timezone)
  }

  const weekdayExpression = value.match(
    /\b(?:lundi|mardi|mercredi|jeudi|vendredi|samedi|dimanche)\b/
  )
  if (weekdayExpression) {
    const weekday = WEEKDAYS[weekdayExpression[0]]
    if (weekday === undefined) return toRange(today, today, raw || null, timezone)
    const weekdayDirection = value.includes('prochain') || value.includes('suivant')
      ? 'future'
      : value.includes('dernier') || value.includes('passe')
        ? 'past'
        : direction
    const date = resolveWeekday(weekday, today, weekdayDirection)
    return toRange(date, date, raw, timezone)
  }

  return toRange(today, today, raw || null, timezone)
}

export function inferDateRangeFromText(
  text: string,
  now = new Date(),
  direction: 'past' | 'future' | 'nearest' = 'nearest',
  timezone = process.env['PAUL_TIMEZONE'] || 'Europe/Paris'
): PaulDateRange | null {
  const normalized = normalize(text)
  const patterns = [
    /aujourd'hui|aujourd hui|ce jour|demain|hier|la veille/,
    /dans\s+\d+\s+jours?/,
    /cette semaine|semaine derniere|semaine prochaine|ce mois(?:-ci)?/,
    /\b\d{1,2}[/-]\d{1,2}(?:[/-]\d{2,4})?\b/,
    /\b(?:lundi|mardi|mercredi|jeudi|vendredi|samedi|dimanche)\b/
  ]
  const matched = patterns.find((pattern) => pattern.test(normalized))
  return matched ? resolveDateRange(normalized.match(matched)?.[0], now, direction, timezone) : null
}

export function formatFrenchDate(day: string): string {
  return new Intl.DateTimeFormat('fr-FR', {
    dateStyle: 'long',
    timeZone: 'UTC'
  }).format(parseDayKey(day))
}

File: Agent-OS-V11/overlay/usr/plugins/agent_os_memory/helpers/temporal.py
from __future__ import annotations

import re
import unicodedata
from datetime import date, datetime, time, timedelta
from zoneinfo import ZoneInfo


WEEKDAYS = {
    "lundi": 0,
    "mardi": 1,
    "mercredi": 2,
    "jeudi": 3,
    "vendredi": 4,
    "samedi": 5,
    "dimanche": 6,
}

MONTHS = {
    "janvier": 1,
    "fevrier": 2,
    "mars": 3,
    "avril": 4,
    "mai": 5,
    "juin": 6,
    "juillet": 7,
    "aout": 8,
    "septembre": 9,
    "octobre": 10,
    "novembre": 11,
    "decembre": 12,
}


def normalize_fr(value: str) -> str:
    text = unicodedata.normalize("NFKD", str(value or "").casefold())
    text = "".join(ch for ch in text if not unicodedata.combining(ch))
    text = text.replace("’", "'")
    text = re.sub(r"[^a-z0-9:/+\-\s']+", " ", text)
    return " ".join(text.split())


def ensure_reference(reference: datetime | None, timezone: str = "Europe/Paris") -> datetime:
    tz = ZoneInfo(timezone)
    current = reference or datetime.now(tz)
    if current.tzinfo is None:
        return current.replace(tzinfo=tz)
    return current.astimezone(tz)


def extract_clock(value: str) -> tuple[int, int] | None:
    text = normalize_fr(value)
    match = re.search(r"(?:^|\b(?:a|vers)\s+)([01]?\d|2[0-3])(?:\s*(?:h|:)\s*([0-5]\d))?\b", text)
    if not match:
        match = re.search(r"\b([01]?\d|2[0-3])\s*h\s*([0-5]\d)?\b", text)
    if not match:
        return None
    return int(match.group(1)), int(match.group(2) or 0)


def _with_clock(day: date, clock: tuple[int, int] | None, tz: ZoneInfo) -> datetime:
    hour, minute = clock or (12, 0)
    return datetime.combine(day, time(hour=hour, minute=minute), tzinfo=tz)


def _resolve_weekday(target: int, reference: datetime, prefer: str, text: str) -> date:
    today = reference.date()
    current = today.weekday()
    normalized = normalize_fr(text)

    if "prochain" in normalized or "prochaine" in normalized:
        delta = (target - current) % 7
        if delta == 0:
            delta = 7
        return today + timedelta(days=delta)

    if "dernier" in normalized or "derniere" in normalized:
        delta = (current - target) % 7
        if delta == 0:
            delta = 7
        return today - timedelta(days=delta)

    if prefer == "past":
        delta = (current - target) % 7
        return today - timedelta(days=delta)

    delta = (target - current) % 7
    return today + timedelta(days=delta)


def resolve_temporal_expression(
    expression: str,
    *,
    reference: datetime | None = None,
    timezone: str = "Europe/Paris",
    prefer: str = "future",
    clock_source: str = "",
) -> datetime | None:
    """Resolve a French natural time expression without semantic guessing.

    ``prefer`` controls an unqualified weekday: ``past`` for completed actions,
    ``future`` for tasks/appointments/events. The model only supplies the
    original expression; calendar arithmetic remains deterministic here.
    """

    reference = ensure_reference(reference, timezone)
    tz = ZoneInfo(timezone)
    raw = str(expression or "").strip()
    text = normalize_fr(raw)
    if not text:
        return None

    clock = extract_clock(f"{raw} {clock_source}".strip())

    # ISO date or datetime.
    iso_match = re.search(r"\b(\d{4}-\d{2}-\d{2})(?:[t\s](\d{1,2}):(\d{2}))?\b", text)
    if iso_match:
        day = date.fromisoformat(iso_match.group(1))
        if iso_match.group(2):
            clock = (int(iso_match.group(2)), int(iso_match.group(3)))
        return _with_clock(day, clock, tz)

    # dd/mm/yyyy or dd-mm-yyyy; two-digit years are accepted as 20xx.
    numeric = re.search(r"\b(0?[1-9]|[12]\d|3[01])[/\-](0?[1-9]|1[0-2])(?:[/\-](\d{2}|\d{4}))?\b", text)
    if numeric:
        day_num = int(numeric.group(1))
        month_num = int(numeric.group(2))
        year_raw = numeric.group(3)
        year = reference.year if not year_raw else int(year_raw)
        if year < 100:
            year += 2000
        candidate = date(year, month_num, day_num)
        if not year_raw:
            if prefer == "future" and candidate < reference.date():
                candidate = date(year + 1, month_num, day_num)
            elif prefer == "past" and candidate > reference.date():
                candidate = date(year - 1, month_num, day_num)
        return _with_clock(candidate, clock, tz)

    # 12 septembre [2026]
    month_names = "|".join(MONTHS)
    named = re.search(rf"\b(0?[1-9]|[12]\d|3[01])\s+({month_names})(?:\s+(\d{{4}}))?\b", text)
    if named:
        year = int(named.group(3) or reference.year)
        candidate = date(year, MONTHS[named.group(2)], int(named.group(1)))
        if not named.group(3):
            if prefer == "future" and candidate < reference.date():
                candidate = date(year + 1, candidate.month, candidate.day)
            elif prefer == "past" and candidate > reference.date():
                candidate = date(year - 1, candidate.month, candidate.day)
        return _with_clock(candidate, clock, tz)

    # Longest expressions first so ``avant hier`` is not swallowed by ``hier``
    # and ``après demain`` is not swallowed by ``demain``.
    relatives = {
        "avant hier": -2,
        "avant-hier": -2,
        "apres demain": 2,
        "apres-demain": 2,
        "aujourd'hui": 0,
        "aujourdhui": 0,
        "aujourd hui": 0,
        "hier": -1,
        "demain": 1,
    }
    for marker, offset in relatives.items():
        if marker in text:
            return _with_clock(reference.date() + timedelta(days=offset), clock, tz)

    for name, weekday in WEEKDAYS.items():
        if re.search(rf"\b{name}\b", text):
            return _with_clock(_resolve_weekday(weekday, reference, prefer, text), clock, tz)

    return None




NUMBER_WORDS = {
    "un": 1, "une": 1, "deux": 2, "trois": 3, "quatre": 4,
    "cinq": 5, "six": 6, "sept": 7, "huit": 8, "neuf": 9,
    "dix": 10, "onze": 11, "douze": 12,
}


def _number_token(value: str) -> int | None:
    token = normalize_fr(value)
    if token.isdigit():
        return int(token)
    return NUMBER_WORDS.get(token)


def _subtract_months(day: date, months: int) -> date:
    index = day.year * 12 + (day.month - 1) - months
    year, month0 = divmod(index, 12)
    month = month0 + 1
    # Clamp the day to the last valid day in the target month.
    if month == 12:
        next_month = date(year + 1, 1, 1)
    else:
        next_month = date(year, month + 1, 1)
    last_day = (next_month - timedelta(days=1)).day
    return date(year, month, min(day.day, last_day))


def resolve_period_expression(
    expression: str,
    *,
    reference: datetime | None = None,
    timezone: str = "Europe/Paris",
) -> tuple[date, date] | None:
    reference = ensure_reference(reference, timezone)
    today = reference.date()
    text = normalize_fr(expression)
    if not text:
        return None

    if "cette semaine" in text:
        start = today - timedelta(days=today.weekday())
        return start, today
    if "semaine derniere" in text or "la semaine derniere" in text:
        current_start = today - timedelta(days=today.weekday())
        end = current_start - timedelta(days=1)
        return end - timedelta(days=6), end
    if "ce mois" in text or "ce mois ci" in text:
        return date(today.year, today.month, 1), today
    if "mois dernier" in text or "le mois dernier" in text:
        start_current = date(today.year, today.month, 1)
        end = start_current - timedelta(days=1)
        return date(end.year, end.month, 1), end
    if "cette annee" in text:
        return date(today.year, 1, 1), today
    if "annee derniere" in text or "l annee derniere" in text:
        return date(today.year - 1, 1, 1), date(today.year - 1, 12, 31)

    match = re.search(
        r"(?:ces?|les?|sur|depuis)\s+([a-z0-9]+)\s+derniers?\s+(jours?|semaines?|mois|annees?)",
        text,
    )
    if not match:
        match = re.search(r"depuis\s+([a-z0-9]+)\s+(jours?|semaines?|mois|annees?)", text)
    if match:
        amount = _number_token(match.group(1))
        unit = match.group(2)
        if not amount or amount < 1 or amount > 120:
            return None
        if unit.startswith("jour"):
            return today - timedelta(days=amount), today
        if unit.startswith("semaine"):
            return today - timedelta(weeks=amount), today
        if unit == "mois":
            return _subtract_months(today, amount), today
        if unit.startswith("annee"):
            try:
                start = today.replace(year=today.year - amount)
            except ValueError:
                start = today.replace(year=today.year - amount, day=28)
            return start, today
    return None

def day_expression_to_iso_bounds(
    expression: str,
    *,
    reference: datetime | None = None,
    timezone: str = "Europe/Paris",
    prefer: str = "past",
) -> tuple[str | None, str | None]:
    period = resolve_period_expression(
        expression, reference=reference, timezone=timezone
    )
    if period is not None:
        return period[0].isoformat(), period[1].isoformat()
    moment = resolve_temporal_expression(
        expression,
        reference=reference,
        timezone=timezone,
        prefer=prefer,
    )
    if moment is None:
        return None, None
    day = moment.date().isoformat()
    return day, day
